File: tauri-app/src-tauri/src/paketquellen.rs
//! Die zwei Volltext-Teile des Pakets: Dokumentation und Domänenquellen.
//!
//! Jede Datei erscheint im Paket genau einmal mit Inhalt: Markdown-Dateien in
//! `Dokumentation.md`, alles andere in der Quellendatei der Domäne.
//! Gegenstück in der CLI: `src/paketQuellen.ts`.

use crate::domaene::{Domaene, ist_dokumentation};
use crate::paketbasis::{hinweisblock, kopf};
use crate::scan::{DateiInfo, ScanErgebnis};
use crate::sprache::code_block_sprache;

/// Eine Datei im Volltext.
fn volltext(datei: &DateiInfo) -> String {
    format!(
        "\n---\n\n### `{}`\n\n**Sprache:** {}  **Zeilen:** {}  **Zeichen:** {}\n\n```{}\n{}\n```\n",
        datei.relativer_pfad,
        datei.sprache,
        datei.zeilen,
        datei.zeichen,
        code_block_sprache(&datei.sprache),
        datei.inhalt
    )
}

/// Alle Markdown-Dateien im Volltext (Doku-Linse).
pub fn dokumentation(scan: &ScanErgebnis) -> String {
    let dokumente: Vec<&DateiInfo> = scan
        .dateien
        .iter()
        .filter(|datei| ist_dokumentation(datei))
        .collect();

    let mut text = kopf(
        "Dokumentation",
        scan,
        &[format!("**Dokumente:** {}", dokumente.len())],
    );
    text.push_str(&hinweisblock(scan));

    if dokumente.is_empty() {
        text.push_str("\nKeine Dokumentationsdateien gefunden.\n");
        return text;
    }

    for datei in &dokumente {
        text.push_str(&format!(
            "\n---\n\n## `{}`\n\n**Sprache:** {}  **Zeilen:** {}  **Zeichen:** {}\n\n```{}\n{}\n```\n",
            datei.relativer_pfad,
            datei.sprache,
            datei.zeilen,
            datei.zeichen,
            code_block_sprache(&datei.sprache),
            datei.inhalt
        ));
    }

    text
}

/// Vollständiger Code einer Domäne (ohne Dokumentationsdateien).
pub fn quellen(scan: &ScanErgebnis, domaene: &Domaene<'_>) -> String {
    let dateien: Vec<&DateiInfo> = domaene
        .dateien
        .iter()
        .copied()
        .filter(|datei| !ist_dokumentation(datei))
        .collect();
    let zeilen: usize = dateien.iter().map(|datei| datei.zeilen).sum();

    let mut text = kopf(
        &format!("Quellen der Domäne `{}`", domaene.name),
        scan,
        &[
            format!("**Datei:** `{}`", crate::domaene::quellen_name(domaene)),
            format!("**Umfang:** {} Dateien · {} Zeilen", dateien.len(), zeilen),
        ],
    );
    text.push_str(&hinweisblock(scan));

    if dateien.is_empty() {
        text.push_str(
            "\nDiese Domäne enthält nur Dokumentationsdateien; sie stehen in `Dokumentation.md`.\n",
        );
        return text;
    }

    text.push_str("\n## Dateiindex\n\n");
    for (index, datei) in dateien.iter().enumerate() {
        text.push_str(&format!(
            "{}. `{}` — {}, {} Zeilen, {} Zeichen\n",
            index + 1,
            datei.relativer_pfad,
            datei.sprache,
            datei.zeilen,
            datei.zeichen
        ));
    }

    for datei in &dateien {
        text.push_str(&volltext(datei));
    }

    text
}

File: src/paketQuellen.ts
/**
 * Die zwei Volltext-Teile des Pakets: Dokumentation und Domänenquellen.
 *
 * Jede Datei erscheint im Paket genau einmal mit Inhalt: Markdown-Dateien in
 * `Dokumentation.md`, alles andere in der Quellendatei der Domäne.
 * Gegenstück in der GUI: `tauri-app/src-tauri/src/paketquellen.rs`.
 */
import { Domaene, istDokumentation } from '@propakt/core';
import { hinweisblock, kopf } from './paketBasis';
import { GescannteDatei } from './scanner';
import { SchemaMeta } from '@propakt/core';
import { codeBlockSprache } from '@propakt/core';

/** Eine Datei im Volltext. */
function volltext(datei: GescannteDatei): string {
  let text = `\n---\n\n### \`${datei.relativerPfad}\`\n\n`;
  text += `**Sprache:** ${datei.sprache}  **Zeilen:** ${datei.zeilen}  **Zeichen:** ${datei.zeichen}\n\n`;
  text += `\`\`\`${codeBlockSprache(datei.sprache)}\n${datei.inhalt}\n\`\`\`\n`;
  return text;
}

/** Alle Markdown-Dateien im Volltext (Doku-Linse). */
export function dokumentation(meta: SchemaMeta, dateien: GescannteDatei[]): string {
  const dokumente = dateien.filter(istDokumentation);

  let text = kopf('Dokumentation', meta, [`**Dokumente:** ${dokumente.length}`]);
  text += hinweisblock(meta);

  if (dokumente.length === 0) {
    text += '\nKeine Dokumentationsdateien gefunden.\n';
    return text;
  }

  for (const datei of dokumente) {
    text += `\n---\n\n## \`${datei.relativerPfad}\`\n\n`;
    text += `**Sprache:** ${datei.sprache}  **Zeilen:** ${datei.zeilen}  **Zeichen:** ${datei.zeichen}\n\n`;
    text += `\`\`\`${codeBlockSprache(datei.sprache)}\n${datei.inhalt}\n\`\`\`\n`;
  }

  return text;
}

/** Vollständiger Code einer Domäne (ohne Dokumentationsdateien). */
export function quellen(meta: SchemaMeta, domaene: Domaene, dateiname: string): string {
  const dateien = domaene.dateien.filter((datei) => !istDokumentation(datei));
  const zeilen = dateien.reduce((summe, datei) => summe + datei.zeilen, 0);

  let text = kopf(`Quellen der Domäne \`${domaene.name}\``, meta, [
    `**Datei:** \`${dateiname}\``,
    `**Umfang:** ${dateien.length} Dateien · ${zeilen} Zeilen`,
  ]);
  text += hinweisblock(meta);

  if (dateien.length === 0) {
    text += '\nDiese Domäne enthält nur Dokumentationsdateien; sie stehen in `Dokumentation.md`.\n';
    return text;
  }

  text += '\n## Dateiindex\n\n';
  dateien.forEach((datei, index) => {
    text += `${index + 1}. \`${datei.relativerPfad}\` — ${datei.sprache}, ${datei.zeilen} Zeilen, ${datei.zeichen} Zeichen\n`;
  });

  for (const datei of dateien) {
    text += volltext(datei);
  }

  return text;
}
